third_party/terraform/utils/provider_test.go.erb

```diff
@@ -96,6 +106,88 @@ func init() {
 	<% end -%>
 }
 
+var lock = &sync.Mutex{}
+
+// Returns a cached config if VCR testing is enabled. This enables us to use a single HTTP transport
```

Comment:
If each test runs once per suite, in what circumstances would this be called multiple times such that we get a cache hit?

Reply:
This whole thing is to replace the `schema.Provider.ConfigureFunc` which calls our `LoadAndValidate` in `config.go`. `LoadAndValidate` constructs the HTTP client and sets base paths, client setup etc, so every time it is called we get a new HTTP client.

`ConfigureFunc` gets called many times within a single test. For a test that has 2 steps, create and import you would see the provider get configured ~4 times, IE for every command in:
```
terraform plan
terraform apply
terraform import
terraform plan
```

This lets us call `ConfigureFunc` once and reuse the HTTP client created in the first call

Reply:
I did some digging, and why this happens is each TestStep seems to call out to the terraform executable individually: https://github.com/hashicorp/terraform-plugin-test/blob/ab13c70d86f9a0a136b33cd9f7eb7c9a9f7d939a/terraform.go#L112

So I believe each call to plan, apply, etc (which can happen a lot within a single test step) will be seen as an individual terraform command and each trigger `ConfigureFunc`

Reply:
thanks! can you put some of this info into a comment so that future curious people can find it easier?